### PR TITLE
feat: connect and sign from LIFF

### LINKED ISSUE #16: Add a condition to display the Wallet selection screen instead of the QR code when launching from LIFF

### PR DESCRIPTION
**Pull request by shiki-tak**
close #16 
- Updated to allow launching of wallet when connecting to wallet from LIFF browser
- WalletProvider now supports v2
- Added a button to sign from LIFF browser after connecting to wallet
- Currently, the PairingMetadata url is set in my test environment, so it will need to be corrected later.

https://github.com/user-attachments/assets/2bc67137-5ef1-4bd0-81cb-1b280018d58d

### CODE REVIEW

File: flutter_bird_app/lib/controller/authentication_service.dart [hunk 1]
```diff
@@ -5,17 +5,17 @@ import 'dart:math' as math;
...
 import 'package:nonce/nonce.dart';
 import 'package:url_launcher/url_launcher.dart';
 import 'package:walletconnect_flutter_v2/walletconnect_flutter_v2.dart';
 
-import '../../model/account.dart';
-import '../../model/wallet_provider.dart';
+import '../model/account.dart';
+import '../model/wallet_provider.dart';
 
-/// Manages the authentication process and communication with crypto wallets
```
Comment (Mdaiki0730) [inline flutter_bird_app/lib/controller/authentication_service.dart:17]: Why did you remove this comment?

File: flutter_bird_app/lib/controller/authentication_service.dart [hunk 2]
```diff
@@ -26,69 +26,117 @@ abstract class AuthenticationService {
...
 class AuthenticationServiceImpl implements AuthenticationService {
-  @override
-  late final List<WalletProvider> availableWallets;
-
+  final bool isInLiff;
   final int operatingChain;
+  final Function() onAuthStatusChanged;
+  WalletProvider? _lastUsedWallet;
+
+  static const String projectId = String.fromEnvironment('WALLET_CONNECT_PROJECT_ID');
```
Comment (Mdaiki0730) [inline flutter_bird_app/lib/controller/authentication_service.dart]: Since we have decided to take that approach in our discussions, would you change it so that it is obtained from `.env`?
https://github.com/Finschia/flutter-bird/pull/18#discussion_r1682041638

File: flutter_bird_app/lib/controller/authentication_service.dart [hunk 3]
```diff
@@ -26,69 +26,117 @@ abstract class AuthenticationService {
...
+  @override
+  String? get webQrData => _webQrData;
+  String? _webQrData;
+
+  WalletProvider? get lastUsedWallet => _lastUsedWallet;
+
+  SessionData? get currentSession => _connector?.sessions.getAll().firstOrNull;
   bool get isConnected => currentSession != null;
 
-  // The data to display in a QR Code for connections on Desktop / Browser.
```
Comment (Mdaiki0730) [inline flutter_bird_app/lib/controller/authentication_service.dart:63]: Why did you remove this comment?

File: flutter_bird_app/lib/view/main_menu_view.dart [hunk 4]
```diff
@@ -104,6 +106,16 @@ class _MainMenuViewState extends State<MainMenuView> with AutomaticKeepAliveClie
         );
       }),
     );
```
Comment (Mdaiki0730) [inline flutter_bird_app/lib/view/main_menu_view.dart]: Would you indent this block?

File: flutter_bird_app/lib/main.dart [hunk 5]
```diff
@@ -1,30 +1,51 @@
...
+    await dotenv.load(fileName: ".env");
+  } 
 
-void main() {
-  dotenv.load(fileName: ".env");
-  runApp(const MyApp());
+  print("Starting application...");
+  const String liffId = String.fromEnvironment('LIFF_ID');
+    
+  final String? os = FlutterLineLiff().os;
```
Comment (Mdaiki0730) [inline flutter_bird_app/lib/main.dart]: Where is this used?

File: flutter_bird_app/lib/main.dart [hunk 6]
```diff
@@ -1,30 +1,51 @@
...
+  WidgetsFlutterBinding.ensureInitialized();
+   if (const String.fromEnvironment('FLUTTER_ENV') != 'production') {
+    await dotenv.load(fileName: ".env");
+  } 
 
-void main() {
-  dotenv.load(fileName: ".env");
-  runApp(const MyApp());
+  print("Starting application...");
+  const String liffId = String.fromEnvironment('LIFF_ID');
```
Comment (Mdaiki0730) [inline flutter_bird_app/lib/main.dart]: Where is this used?

File: flutter_bird_app/lib/controller/authentication_service.dart [hunk 7]
```diff
@@ -103,37 +151,70 @@ class AuthenticationServiceImpl implements AuthenticationService {
 
         Uri? uri = resp.uri;
         if (uri != null) {
-          if (kIsWeb) {
-            webQrData = uri.toString();
+          // Web
+          if (kIsWeb && !isInLiff) {
```
Comment (Mdaiki0730) [inline flutter_bird_app/lib/controller/authentication_service.dart:157]: Will this condition also be applied when searching the web on mobile?
If so, how about launching a wallet for every mobile?

File: flutter_bird_app/lib/controller/authentication_service.dart [hunk 8]
```diff
@@ -26,69 +26,117 @@ abstract class AuthenticationService {
...
-  /// Loads all WalletConnect compatible wallets
-  _loadWallets() async {
-    final walletResponse = await http.get(Uri.parse('https://registry.walletconnect.org/data/wallets.json'));
-    final walletData = json.decode(walletResponse.body);
-    availableWallets = walletData.entries.map<WalletProvider>((data) => WalletProvider.fromJson(data.value)).toList();
+  // Update to support WalletConnect v2
+  Future<void> _loadWallets() async {
+    try {
+      final response = await http.get(
+        Uri.parse('https://explorer-api.walletconnect.com/v3/wallets?projectId=${projectId}&entries=5&page=1'),
```
Comment (Mdaiki0730) [inline flutter_bird_app/lib/controller/authentication_service.dart:107]: Does the change in the WalletProvider model correspond to the update in this URL?

File: flutter_bird_app/lib/controller/authentication_service.dart (hunk 7 above)
Comment (shiki-tak) [reply flutter_bird_app/lib/controller/authentication_service.dart:157]: It seems that kIsWeb is true whether you open it from a PC browser or a mobile browser.
Therefore, if I set this condition to launchWallet, it seems that I will not be able to open it from a PC browser.

If I want to separate the behavior for PC browsers and mobile browsers, I will need a different check. So I would like to make that a separate issue. What do you think?

File: flutter_bird_app/lib/main.dart (hunk 6 above)
Comment (shiki-tak) [reply flutter_bird_app/lib/main.dart]: This variable was used previously, but is no longer needed. Delete. Thanks!

File: flutter_bird_app/lib/main.dart (hunk 5 above)
Comment (shiki-tak) [reply flutter_bird_app/lib/main.dart]: This variable was used previously, but is no longer needed. Delete. Thanks!

File: flutter_bird_app/lib/view/main_menu_view.dart (hunk 4 above)
Comment (shiki-tak) [reply flutter_bird_app/lib/view/main_menu_view.dart]: Thanks!

File: flutter_bird_app/lib/controller/authentication_service.dart (hunk 8 above)
Comment (shiki-tak) [reply flutter_bird_app/lib/controller/authentication_service.dart:107]: I've updated the WallertProvider model to a format that can be retrieved by this URL. Does this answer your question?

File: flutter_bird_app/lib/controller/authentication_service.dart (hunk 3 above)
Comment (shiki-tak) [reply flutter_bird_app/lib/controller/authentication_service.dart:63]: There was no need to delete it, but it seems to have been deleted in the middle of coding. I'll put it back.

File: flutter_bird_app/lib/controller/authentication_service.dart (hunk 1 above)
Comment (shiki-tak) [reply flutter_bird_app/lib/controller/authentication_service.dart:17]: There was no need to delete it, but it seems to have been deleted in the middle of coding. I'll put it back.

**Comment by shiki-tak** [comment]
Since ci fails for PRs from forked branches, we will remake them into the following PRs.
https://github.com/Finschia/flutter-bird/pull/19

File: flutter_bird_app/lib/controller/authentication_service.dart (hunk 8 above)
Comment (Mdaiki0730) [reply flutter_bird_app/lib/controller/authentication_service.dart:107]: Yes, it does. Thank you!

File: flutter_bird_app/lib/controller/authentication_service.dart (hunk 7 above)
Comment (Mdaiki0730) [reply flutter_bird_app/lib/controller/authentication_service.dart:157]: Yes, what I meant was to add a check for mobile. 

> If I want to separate the behavior for PC browsers and mobile browsers, I will need a different check. So I would like to make that a separate issue. What do you think?

That sounds good to me.

Comment (shiki-tak) [reply flutter_bird_app/lib/controller/authentication_service.dart:157]: Created: https://github.com/Finschia/flutter-bird/issues/20
